src/main/java/org/takes/rq/RqForm.java

```diff
@@ -156,12 +162,16 @@ private static String decode(final CharSequence txt) {
          * @throws IOException If something fails reading or parsing body
          */
         private Map<String, List<String>> map() throws IOException {
-            synchronized (this.saved) {
-                if (this.saved.isEmpty()) {
-                    this.saved.add(this.freshMap());
+            if (this.saved.get() == null) {
+                synchronized (this.mutex) {
+                    if (this.saved.get() == null) {
```

Comment:
@yegor256 You can dislike my approach but I don't understand why it is wrong, it is thread safe and does the exact same thing with more appropriate objects IMHO. It is the well known double checking approach based on a `volatile` variable, I know that it is a controversy approach but it is not wrong